src/crimson/os/seastore/seastore_types.h

```diff
@@ -1228,39 +1666,102 @@ constexpr laddr_t L_ADDR_NULL = L_ADDR_MAX;
 constexpr laddr_t L_ADDR_ROOT = laddr_t::from_raw_uint(laddr_t::RAW_VALUE_MAX - 1);
 constexpr laddr_t L_ADDR_LBAT = laddr_t::from_raw_uint(laddr_t::RAW_VALUE_MAX - 2);
 
+struct laddr_printer_t {
+  const laddr_t &addr;
+};
+std::ostream &operator<<(std::ostream &, const laddr_printer_t &);
+
+enum class laddr_conflict_level_t {
+  object_prefix,
+  clone_prefix,
+  all,
+};
+
+// see BtreeLBAManager::search_insert_pos(), linear will try the
+// next iterator, and random will try gen_next_hint(hint)
+enum class laddr_conflict_policy_t {
+  linear_search,
+  gen_random,
+  never
+};
+
+struct laddr_hint_t {
+  laddr_t addr = L_ADDR_NULL;
+  laddr_conflict_level_t conflict_level =
+      laddr_conflict_level_t::all;
+  laddr_conflict_policy_t conflict_policy =
+      laddr_conflict_policy_t::never;
+
+  static constexpr laddr_hint_t never_conflict(laddr_t addr) {
+    laddr_hint_t hint{};
+    hint.addr = addr;
+    return hint;
+  }
+
+  bool conflict_with(laddr_t other) const {
+    switch (conflict_level) {
+    case laddr_conflict_level_t::object_prefix:
+      assert(addr.is_object_address());
+      return addr.get_object_prefix() == other.get_object_prefix();
+    case laddr_conflict_level_t::clone_prefix:
+      assert(addr.is_object_address());
+      return addr.get_clone_prefix() == other.get_clone_prefix();
+    case laddr_conflict_level_t::all:
+      return addr == other;
+    default:
+      __builtin_unreachable();
+    }
+  }
+};
+std::ostream &operator<<(std::ostream &out, const laddr_hint_t &hint);
+
+laddr_hint_t gen_global_random_hint();
+laddr_hint_t gen_meta_onode_hint(
+  shard_t shard,
+  pool_t pool,
+  crush_hash_t crush);
+laddr_hint_t gen_object_hint(
+  shard_t shard,
+  pool_t pool,
+  crush_hash_t crush,
+  std::optional<local_object_id_t> object_id,
+  std::optional<local_clone_id_t> clone_id,
+  bool is_metadata);
```

Comment:
Named factory methods might be easier to follow instead of raw parameters -- many combinations are not meaningful.

For example:
`laddr_hint_t::never_conflict(laddr)` -> `laddr_hint_t::create_as_fixed(laddr)`
`gen_global_random_hint()` -> `laddr_hint_t::create_global_md_hint()`
`gen_meta_onode_hint()` -> `laddr_hint_t::create_onode_hint(shard, pool, hash)`
`gen_object_hint()` -> `laddr_hint_t::create_object_hint(shard, pool, hash)`,
-> `laddr_hint_t::create_object_md_hint(shard, pool, hash, object_id)`,
-> `laddr_hint_t::create_object_data_hint_cloned(shard, pool, hash, object_id, clone_id)`
-> `laddr_hint_t::create_test_hint() ?`
`gen_next_hint()` -> `last_hint.find_random_next()`

Not sure if I missed any use case.

Reply:
I will refactor the laddr_hint_t related comments and repush in the next few days.

Reply:
I'll keep looking at the reset.